File: Services/SystemRestoreService.cs
using System;
using System.Diagnostics;
using System.Threading.Tasks;

namespace InvolveX.Cli.Services
{
    public class SystemRestoreService
    {
        private readonly LogService _logService;

        public SystemRestoreService(LogService logService)
        {
            _logService = logService;
        }

        public async Task<bool> CreateRestorePoint(string description = "InvolveX CLI Manual Restore Point")
        {
            try
            {
                _logService.Log($"Creating system restore point: {description}");

                var process = new Process
                {
                    StartInfo = new ProcessStartInfo
                    {
                        FileName = "powershell",
                        Arguments = $"-Command \"Checkpoint-Computer -Description '{description}' -RestorePointType 'MODIFY_SETTINGS'\"",
                        RedirectStandardOutput = true,
                        RedirectStandardError = true,
                        UseShellExecute = false,
                        CreateNoWindow = true,
                        Verb = "runas" // Request admin privileges
                    }
                };

                process.Start();

                var outputTask = process.StandardOutput.ReadToEndAsync();
                var errorTask = process.StandardError.ReadToEndAsync();

                await Task.WhenAll(outputTask, errorTask);
                await process.WaitForExitAsync();

                var output = outputTask.Result;
                var error = errorTask.Result;

                _logService.Log($"[Create Restore Point STDOUT]: {output}");
                _logService.Log($"[Create Restore Point STDERR]: {error}");

                if (process.ExitCode == 0)
                {
                    _logService.Log("System restore point created successfully");
                    return true;
                }
                else
                {
                    _logService.Log($"Failed to create restore point. Exit code: {process.ExitCode}");
                    return false;
                }
            }
            catch (Exception ex)
            {
                _logService.Log($"Exception creating restore point: {ex.Message}");
                return false;
            }
        }

        public async Task<string> ListRestorePoints()
        {
            try
            {
                _logService.Log("Listing system restore points");

                var process = new Process
                {
                    StartInfo = new ProcessStartInfo
                    {
                        FileName = "powershell",
                        Arguments = "-Command \"Get-ComputerRestorePoint | Select-Object -Property SequenceNumber, Description, CreationTime, RestorePointType | Sort-Object -Property CreationTime -Descending | Format-Table -AutoSize\"",
                        RedirectStandardOutput = true,
                        RedirectStandardError = true,
                        UseShellExecute = false,
                        CreateNoWindow = true
                    }
                };

                process.Start();

                var outputTask = process.StandardOutput.ReadToEndAsync();
                var errorTask = process.StandardError.ReadToEndAsync();

                await Task.WhenAll(outputTask, errorTask);
                await process.WaitForExitAsync();

                var output = outputTask.Result;
                var error = errorTask.Result;

                _logService.Log($"[List Restore Points STDOUT]: {output}");
                _logService.Log($"[List Restore Points STDERR]: {error}");

                if (process.ExitCode == 0)
                {
                    return string.IsNullOrWhiteSpace(output) ? "No restore points found." : output;
                }
                else
                {
                    return $"Error listing restore points: {error}";
                }
            }
            catch (Exception ex)
            {
                _logService.Log($"Exception listing restore points: {ex.Message}");
                return $"Error: {ex.Message}";
            }
        }

        public async Task<bool> DeleteRestorePoint(int sequenceNumber)
        {
            try
            {
                _logService.Log($"Deleting restore point with sequence number: {sequenceNumber}");

                var process = new Process
                {
                    StartInfo = new ProcessStartInfo
                    {
                        FileName = "powershell",
                        Arguments = $"-Command \"Remove-ComputerRestorePoint -SequenceNumber {sequenceNumber} -Confirm:$false\"",
                        RedirectStandardOutput = true,
                        RedirectStandardError = true,
                        UseShellExecute = false,
                        CreateNoWindow = true,
                        Verb = "runas" // Request admin privileges
                    }
                };

                process.Start();

                var outputTask = process.StandardOutput.ReadToEndAsync();
                var errorTask = process.StandardError.ReadToEndAsync();

                await Task.WhenAll(outputTask, errorTask);
                await process.WaitForExitAsync();

                var output = outputTask.Result;
                var error = errorTask.Result;

                _logService.Log($"[Delete Restore Point STDOUT]: {output}");
                _logService.Log($"[Delete Restore Point STDERR]: {error}");

                if (process.ExitCode == 0)
                {
                    _logService.Log("System restore point deleted successfully");
                    return true;
                }
                else
                {
                    _logService.Log($"Failed to delete restore point. Exit code: {process.ExitCode}");
                    return false;
                }
            }
            catch (Exception ex)
            {
                _logService.Log($"Exception deleting restore point: {ex.Message}");
                return false;
            }
        }

        public async Task<bool> DeleteOldRestorePoints(int keepCount = 5)
        {
            try
            {
                _logService.Log($"Deleting old restore points, keeping {keepCount} most recent");

                // First get all restore points
                var listProcess = new Process
                {
                    StartInfo = new ProcessStartInfo
                    {
                        FileName = "powershell",
                        Arguments = "-Command \"Get-ComputerRestorePoint | Select-Object -Property SequenceNumber | Sort-Object -Property SequenceNumber -Descending\"",
                        RedirectStandardOutput = true,
                        RedirectStandardError = true,
                        UseShellExecute = false,
                        CreateNoWindow = true
                    }
                };

                listProcess.Start();

                var listOutputTask = listProcess.StandardOutput.ReadToEndAsync();
                var listErrorTask = listProcess.StandardError.ReadToEndAsync();

                await Task.WhenAll(listOutputTask, listErrorTask);
                await listProcess.WaitForExitAsync();

                if (listProcess.ExitCode != 0)
                {
                    _logService.Log("Failed to list restore points for cleanup");
                    return false;
                }

                var output = listOutputTask.Result;
                var sequenceNumbers = new System.Collections.Generic.List<int>();

                // Parse sequence numbers from output
                foreach (var line in output.Split('\n'))
                {
                    if (int.TryParse(line.Trim(), out int seqNum))
                    {
                        sequenceNumbers.Add(seqNum);
                    }
                }

                if (sequenceNumbers.Count <= keepCount)
                {
                    _logService.Log($"Only {sequenceNumbers.Count} restore points found, no cleanup needed");
                    return true;
                }

                // Delete old restore points (keep the most recent ones)
                var deletedCount = 0;
                for (int i = keepCount; i < sequenceNumbers.Count; i++)
                {
                    var success = await DeleteRestorePoint(sequenceNumbers[i]);
                    if (success)
                    {
                        deletedCount++;
                    }
                }

                _logService.Log($"Deleted {deletedCount} old restore points");
                return true;
            }
            catch (Exception ex)
            {
                _logService.Log($"Exception cleaning up restore points: {ex.Message}");
                return false;
            }
        }

        public async Task<bool> IsSystemRestoreEnabled()
        {
            try
            {
                var process = new Process
                {
                    StartInfo = new ProcessStartInfo
                    {
                        FileName = "powershell",
                        Arguments = "-Command \"(Get-ComputerRestorePoint).Count -gt 0\"",
                        RedirectStandardOutput = true,
                        RedirectStandardError = true,
                        UseShellExecute = false,
                        CreateNoWindow = true
                    }
                };

                process.Start();

                var outputTask = process.StandardOutput.ReadToEndAsync();
                var errorTask = process.StandardError.ReadToEndAsync();

                await Task.WhenAll(outputTask, errorTask);
                await process.WaitForExitAsync();

                var output = outputTask.Result.Trim();

                return process.ExitCode == 0 && output == "True";
            }
            catch (Exception ex)
            {
                _logService.Log($"Exception checking system restore status: {ex.Message}");
                return false;
            }
        }
    }
}

File: services/SystemRestoreService.js
const { spawn } = require('child_process');

class SystemRestoreService {
  constructor(logService) {
    this.logService = logService;
  }

  async createRestorePoint(description) {
    try {
      this.logService.log(`Creating system restore point: ${description}`);

      const psScript = `
                # Create a system restore point
                try {
                    $description = "${description}"
                    $restorePoint = Get-ComputerRestorePoint | Where-Object { $_.Description -eq $description }

                    if ($restorePoint) {
                        Write-Output "Restore point already exists"
                        exit 1
                    }

                    Checkpoint-Computer -Description $description -RestorePointType "MODIFY_SETTINGS"
                    Write-Output "SUCCESS"
                } catch {
                    Write-Output "ERROR: $($_.Exception.Message)"
                    exit 1
                }
            `;

      const result = await this.runProcess('powershell', ['-Command', psScript]);

      if (result.stdout.includes('SUCCESS')) {
        this.logService.log('System restore point created successfully');
        return true;
      } else {
        this.logService.log(`Failed to create restore point: ${result.stdout}`);
        return false;
      }
    } catch (error) {
      this.logService.log(`Error creating restore point: ${error.message}`);
      return false;
    }
  }

  async listRestorePoints() {
    try {
      this.logService.log('Listing system restore points');

      const psScript = `
                try {
                    $restorePoints = Get-ComputerRestorePoint | Sort-Object -Property CreationTime -Descending
                    if ($restorePoints.Count -eq 0) {
                        Write-Output "No restore points found"
                    } else {
                        foreach ($point in $restorePoints) {
                            Write-Output "Sequence: $($point.SequenceNumber) - $($point.Description) - $($point.CreationTime)"
                        }
                    }
                } catch {
                    Write-Output "Error listing restore points: $($_.Exception.Message)"
                }
            `;

      const result = await this.runProcess('powershell', ['-Command', psScript]);
      return result.stdout.trim();
    } catch (error) {
      this.logService.log(`Error listing restore points: ${error.message}`);
      return `Error: ${error.message}`;
    }
  }

  async deleteRestorePoint(sequenceNumber) {
    try {
      this.logService.log(`Deleting restore point with sequence: ${sequenceNumber}`);

      const psScript = `
                try {
                    $sequence = ${sequenceNumber}
                    $restorePoint = Get-ComputerRestorePoint | Where-Object { $_.SequenceNumber -eq $sequence }

                    if (-not $restorePoint) {
                        Write-Output "NOT_FOUND"
                        exit 1
                    }

                    # Note: PowerShell doesn't have a direct way to delete individual restore points
                    # This would require using vssadmin or similar tools
                    Write-Output "DELETE_NOT_SUPPORTED"
                    exit 1
                } catch {
                    Write-Output "ERROR: $($_.Exception.Message)"
                    exit 1
                }
            `;

      const result = await this.runProcess('powershell', ['-Command', psScript]);

      if (result.stdout.includes('NOT_FOUND')) {
        this.logService.log('Restore point not found');
        return false;
      } else if (result.stdout.includes('DELETE_NOT_SUPPORTED')) {
        this.logService.log('Individual restore point deletion not supported');
        return false;
      } else {
        this.logService.log('Restore point deleted successfully');
        return true;
      }
    } catch (error) {
      this.logService.log(`Error deleting restore point: ${error.message}`);
      return false;
    }
  }

  async deleteOldRestorePoints(keepCount) {
    try {
      this.logService.log(`Deleting old restore points, keeping ${keepCount} most recent`);

      const psScript = `
                try {
                    $keepCount = ${keepCount}
                    $restorePoints = Get-ComputerRestorePoint | Sort-Object -Property CreationTime -Descending

                    if ($restorePoints.Count -le $keepCount) {
                        Write-Output "No old restore points to delete"
                        exit 0
                    }

                    # Note: This is a simplified implementation
                    # In practice, you'd need to use vssadmin or similar tools to delete old restore points
                    Write-Output "DELETE_OLD_NOT_SUPPORTED"
                    exit 1
                } catch {
                    Write-Output "ERROR: $($_.Exception.Message)"
                    exit 1
                }
            `;

      const result = await this.runProcess('powershell', ['-Command', psScript]);

      if (result.stdout.includes('No old restore points')) {
        this.logService.log('No old restore points to delete');
        return true;
      } else if (result.stdout.includes('DELETE_OLD_NOT_SUPPORTED')) {
        this.logService.log('Deleting old restore points not supported in this implementation');
        return false;
      } else {
        this.logService.log('Old restore points deleted successfully');
        return true;
      }
    } catch (error) {
      this.logService.log(`Error deleting old restore points: ${error.message}`);
      return false;
    }
  }

  async runProcess(command, args) {
    return new Promise((resolve, reject) => {
      const process = spawn(command, args, {
        stdio: ['pipe', 'pipe', 'pipe'],
        shell: true,
      });

      let stdout = '';
      let stderr = '';

      process.stdout.on('data', data => {
        stdout += data.toString();
      });

      process.stderr.on('data', data => {
        stderr += data.toString();
      });

      process.on('close', code => {
        resolve({
          code,
          stdout,
          stderr,
        });
      });

      process.on('error', error => {
        reject(error);
      });
    });
  }
}

module.exports = SystemRestoreService;
